src/VisualStudio/Next/Remote/ServiceHubRemoteHostClient.JsonRpcSnapshot.cs

```diff
@@ -0,0 +1,159 @@
+            }
+
+            protected override void OnDisposed()
+            {
+                // dispose cancellation registration
+                _cancellationRegistration.Dispose();
+
+                // dispose service and snapshot channels
+                _serviceClient.Dispose();
+                _snapshotClient.Dispose();
+            }
+
+            private class ServiceJsonRpcClient : JsonRpcClient
+            {
+                private readonly object _callbackTarget;
+
+                public ServiceJsonRpcClient(Stream stream, object callbackTarget) : base(stream)
+                {
+                    // this one doesn't need cancellation token since it has nothing to cancel
+                    _callbackTarget = callbackTarget;
+                }
+
+                protected override object GetTarget() => _callbackTarget;
+            }
+
+            private class SnapshotJsonRpcClient : JsonRpcClient
+            {
+                private readonly JsonRpcSession _owner;
+                private readonly CancellationTokenSource _source;
+
+                public SnapshotJsonRpcClient(JsonRpcSession owner, Stream stream) :
+                    base(stream)
+                {
+                    _owner = owner;
+                    _source = new CancellationTokenSource();
+                }
+
+                private ChecksumScope ChecksumScope => _owner.ChecksumScope;
+                protected override object GetTarget() => this;
+
+                public async Task RequestAssetAsync(int serviceId, byte[] checksum, string streamName)
+                {
+                    try
+                    {
+                        // this is callback from remote host side to get asset associated with checksum from VS.
+                        var service = ChecksumScope.Workspace.Services.GetRequiredService<ISolutionChecksumService>();
+
+                        using (var stream = new ClientDirectStream(streamName))
+                        {
+                            await stream.ConnectAsync(_source.Token).ConfigureAwait(false);
+
+                            using (var writer = new ObjectWriter(stream))
+                            {
+                                writer.WriteInt32(serviceId);
+                                writer.WriteArray(checksum);
+
+                                var checksumObject = service.GetChecksumObject(new Checksum(checksum), _source.Token);
+                                writer.WriteString(checksumObject.Kind);
+
+                                await checksumObject.WriteToAsync(writer, _source.Token).ConfigureAwait(false);
+                            }
+
+                            await stream.FlushAsync(_source.Token).ConfigureAwait(false);
+
+                            // TODO: think of a way this is not needed
+                            // wait for the other side to finish reading data I sent over
+                            stream.WaitForServer();
```

Comment:
ya, I made that way in last checkin. see remote host one you have put all other feedback.